backend/src/apiserver/server/util.go

```diff
@@ -241,3 +244,83 @@ func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.P
 	}
 	return nil
 }
+
+
+// convertNodeID converts an old nodeID to a new nodeID
+func convertNodeID(newWf *wfv1.Workflow, regex *regexp.Regexp, oldNodeID string, oldNodes map[string]wfv1.NodeStatus) string {
+	node := oldNodes[oldNodeID]
+	newNodeName := regex.ReplaceAllString(node.Name, newWf.ObjectMeta.Name)
+	return newWf.NodeID(newNodeName)
+}
+
+// FormulateResubmitWorkflow formulate a new workflow from a previous workflow optionally re-using successful nodes
+func formulateResubmitWorkflow(wf *wfv1.Workflow, randomString util.RandomStringInterface) (*wfv1.Workflow, error) {
+	newWF := wfv1.Workflow{}
+	newWF.TypeMeta = wf.TypeMeta
+
+	// Resubmitted workflow will use generated names
+	if wf.ObjectMeta.GenerateName != "" {
+		newWF.ObjectMeta.GenerateName = wf.ObjectMeta.GenerateName
+	} else {
+		newWF.ObjectMeta.GenerateName = wf.ObjectMeta.Name + "-"
+	}
+	// When resubmitting workflow with memoized nodes, we need to use a predetermined workflow name
+	// in order to formulate the node statuses. Which means we cannot reuse metadata.generateName
+	// The following simulates the behavior of generateName
+	newWF.ObjectMeta.Name = newWF.ObjectMeta.GenerateName + randomString.Get(5)
+
+	// carry over the unmodified spec
+	newWF.Spec = wf.Spec
+
+	// carry over user labels and annotations from previous workflow.
+	// skip any argoproj.io labels except for the controller instanceID label.
+	for key, val := range wf.ObjectMeta.Labels {
+		if strings.HasPrefix(key, workflow.FullName+"/") && key != workflow.FullName+"/controller-instanceid" {
+			continue
+		}
+		if newWF.ObjectMeta.Labels == nil {
+			newWF.ObjectMeta.Labels = make(map[string]string)
+		}
+		newWF.ObjectMeta.Labels[key] = val
+	}
+	for key, val := range wf.ObjectMeta.Annotations {
+		if newWF.ObjectMeta.Annotations == nil {
```

Comment:
Is it same as newWF.ObjectMetadata.Annotations = wf.ObjectMetadata.Annotations?

Reply:
here we might want to copy the value instead of the ref.
https://stackoverflow.com/questions/23057785/how-to-copy-a-map/23058707#23058707